pkg/csr/csr.go

```diff
@@ -0,0 +1,357 @@
+
+// kubeletClientUsages contains the permitted key usages from a kube-apiserver-client-kubelet signer
+var kubeletClientUsages = []certificates.KeyUsage{
+	certificates.UsageKeyEncipherment,
+	certificates.UsageDigitalSignature,
+	certificates.UsageClientAuth,
+}
+
+// kubeletServerUsages contains the permitted key usages from a kubelet-serving signer
+var kubeletServerUsages = []certificates.KeyUsage{
+	certificates.UsageKeyEncipherment,
+	certificates.UsageDigitalSignature,
+	certificates.UsageServerAuth,
+}
+
+const (
+	nodeGroup          = "system:nodes"
+	nodeUserName       = "system:node"
+	NodeUserNamePrefix = nodeUserName + ":"
+	systemPrefix       = "system:authenticated"
+)
+
+// Approver holds the information required to approve a node CSR
+type Approver struct {
+	// client is the cache client
+	client client.Client
+	// k8sclientset holds the kube client needed for updating CSR approval conditions
+	k8sclientset *kubernetes.Clientset
+	// csr holds the pointer to the CSR request
+	csr      *certificates.CertificateSigningRequest
+	log      logr.Logger
+	recorder record.EventRecorder
+	// namespace is the namespace in which CSR's are present
+	namespace string
+}
+
+// NewApprover returns a pointer to the Approver
+func NewApprover(client client.Client, clientSet *kubernetes.Clientset, csr *certificates.CertificateSigningRequest,
+	log logr.Logger, recorder record.EventRecorder, watchNamespace string) (*Approver, error) {
+	if client == nil || csr == nil || clientSet == nil {
+		return nil, errors.New("kubernetes client, clientSet or CSR should not be nil")
+	}
+	return &Approver{client,
+		clientSet,
+		csr,
+		log,
+		recorder,
+		watchNamespace}, nil
+}
+
+// Approve approves a CSR by updating it's status conditions to true if it is a valid CSR
+func (a *Approver) Approve() error {
+	if a.k8sclientset == nil {
+		return errors.New("Kubernetes clientSet should not be nil")
+	}
+
+	if validated, err := a.validateCSRContents(); !validated {
+		a.log.Info("CSR contents are invalid for approval by WMCO CSR Approver", "CSR", a.csr.Name)
+		return errors.Wrapf(err, "could not validate CSR %s contents for approval", a.csr.Name)
+	}
+
+	a.csr.Status.Conditions = append(a.csr.Status.Conditions, certificates.CertificateSigningRequestCondition{
+		Type:           certificates.CertificateApproved,
+		Status:         "True",
+		Message:        "Approved by WMCO",
+		LastUpdateTime: meta.Now(),
+		Reason:         "WMCOApproved",
+	})
+
+	if _, err := a.k8sclientset.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.Background(),
+		a.csr.Name, a.csr, meta.UpdateOptions{}); err != nil {
+		return errors.Wrapf(err, "could not update conditions for CSR approval %s", a.csr.Name)
+	}
+	a.log.Info("CSR approved", "CSR", a.csr.Name)
+	return nil
+}
+
+// validateCSRContents returns true if the CSR request contents are valid
+func (a *Approver) validateCSRContents() (bool, error) {
```

Comment:
Whats the point in the bool return type if it only ever returns
false + error
or
true + no error?
error always implies false and no error implies true

Reply:
It was returning false + nil from the validateNodeName() call if the match wasn't found
Realized that could be confusing to read, hence modified the function to return false, nil if node name validation fails with no error.

Reply:
The call site of this function looks strange as we are not checking the error there. I wish we could be more explicit about it. But I will let it go for now.